services/placement/filter/mirrored.go

```diff
@@ -0,0 +1,435 @@
+		pairs := make([]host, 0, rf)
+		for i := 0; i < rf; i++ {
+			var r *rack
+			for rh.Len() > 0 {
+				// Always prefer to take from the largest rack.
+				r = rh.Pop().(*rack)
+				if _, ok := seenRacks[r.rack]; !ok {
+					seenRacks[r.rack] = r
+					break
+				}
+			}
+			if r == nil {
+				break
+			}
+			// Move the host from the rack to the pair.
+			pairs = append(pairs, r.hosts[len(r.hosts)-1])
+			r.hosts = r.hosts[:len(r.hosts)-1]
+		}
+		if len(pairs) == rf {
+			res = append(res, pairs)
+		}
+		for _, r := range seenRacks {
+			if len(r.hosts) > 0 {
+				rh.Push(r)
+			}
+		}
+	}
+
+	return res
+}
+
+func pairInstances(hostPairs [][]host) ([][]services.PlacementInstance, error) {
+	var instancePairs = make([][]services.PlacementInstance, 0, len(hostPairs))
+	for _, hostPair := range hostPairs {
+		for port, instance := range hostPair[0].portToInstance {
+			instancePair := make([]services.PlacementInstance, 0, len(hostPair))
+			instancePair = append(instancePair, instance)
+			for _, otherHost := range hostPair[1:] {
+				otherInstance, ok := otherHost.portToInstance[port]
+				if !ok {
+					return nil, fmt.Errorf("could not find port %s on host %s", port, otherHost.name)
+				}
+				instancePair = append(instancePair, otherInstance)
+			}
+			instancePairs = append(instancePairs, instancePair)
+		}
+	}
+	return instancePairs, nil
+}
+
+func assignShardsToPairedInstances(
+	pairs [][]services.PlacementInstance,
+	p services.Placement,
+) []services.PlacementInstance {
+	var (
+		instances   = make([]services.PlacementInstance, 0, p.ReplicaFactor()*len(pairs))
+		shardSetIDs = nextNShardSetIDs(p, len(pairs))
+	)
+	for i, pair := range pairs {
+		ssID := shardSetIDs[i]
+		for _, instance := range pair {
+			instances = append(instances, instance.SetShardSetID(ssID))
+		}
+	}
+
+	return instances
+}
+
+func nextNShardSetIDs(p services.Placement, n int) []string {
+	var (
+		instances          = p.Instances()
+		currentShardSetIDs = make(map[int]struct{}, len(instances))
+	)
+	for _, instance := range instances {
+		ssID, err := strconv.Atoi(instance.ShardSetID())
+		if err != nil {
+			// NB(cw): the ShardSetID field in instance is a string, so not neccessarily
+			// always convertable to int.
+			// For this function, we can safely ignore non-integer shard set ids,
+			// we just need to find next n smallest integer shard set ids.
+			continue
+		}
+		currentShardSetIDs[ssID] = struct{}{}
+	}
+
+	var (
+		result   = make([]string, 0, n)
+		curIndex = -1
+	)
+	// Find the next n smallest integers to be used as shard set ids.
+	for len(result) < n {
+		curIndex++
+		if _, ok := currentShardSetIDs[curIndex]; ok {
+			continue
+		}
+		result = append(result, strconv.Itoa(curIndex))
+	}
+
+	return result
+}
+
+type host struct {
+	name           string
+	rack           string
+	weight         uint32
+	portToInstance map[string]services.PlacementInstance
+}
+
+type rack struct {
+	rack  string
+	hosts []host
+}
+
+type rackHeap struct {
```

Comment:
The slight inconvenience of dereferencing pointers is probably worth the effort. It saves you some typing too. FWIW this is the example in the `container/heap` package: https://play.golang.org/p/Vz_mpb85uR.